RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+The proposed changes must either be backwards compatible, or be segregated so that Template Users can progressively opt in to version 2.0
+
+## Proposed Implementation Design
+
+### Opt-in instead of opt-out
+
+Currently the templating will apply everything, and Template Users must opt-out.  The new templating system will use a composition model, that is, Template User must opt-in to things.
+
+### Easier to discover, and validate, template settings
+
+Currently Template User need to read the README to determine what settings are available or even valid. This also assumes the README is even correct.  The new templating system will have a metadata, and optionally, a validation layer so that users can discover what templates are available to opt-in to, and optionally, to validate template settings.
+
+### Easier to manage template file locations
+
+The new templating design will have descriptive and well documented file locations
+
+
+**V1 locations**
+
+| Path                      | Description |
+| ------------------------- | ----------- |
+| `/moduleroot`             | Templates get deployed on new module, convert and update |
+| `/moduleroot_init`        | Templates get only deployed when the target module does not yet exist |
+| `/object_templates/*.erb` | Template location for `pdk new ...` generation |
+| `/robocop`                | Custom ruby code to assist in generating rubocop profiles |
+
+**V2 locations**
+
+Where `name` is a unique name for the template
+
+> The `name` should use snake case (lowercase with underscore) to be cross platform compliant. The more descriptive name can be specified in the template.yml file
+
+| Path                               | Description |
+| ---------------------------------- | ----------- |
+| `/<name>/README.md`            (1) | A plain text description of the template. Though not really necessary as most of this would be contained in the `template.yml` file. It is expected that this file will be auto-generated |
+| `/<name>/template.yml`             | YAML file describing the template |
+| `/<name>/template_schema.json` (1) | JSON Schema file describing the valid settings for this template |
+| `/<name>/files/**/*`               | Templates to process for new module, convert and update |
+| `/<name>/files_init/**/*`      (1) | Templates to process when the module does not yet exist |
+| `/<name>/*`                  (1,2) | Additional files used _for_ templating, for example configuration information |
+
+(1) Optional
+
+(2) Ruby files that will be evaluated for the ERB templating process **must** be specified in the `template.yml` file. This is to help (somewhat) safeguard arbitrary Ruby being evaluated. Although it can be argued that ERB templates are also arbitrary code.
+
+> TODO: Does files_init EVEN make sense?!??!
+> Actually it does it in an `always_apply` template
+
+> TODO: What about object_templates?
+
+Note - `/object_templates/*.erb` will still be consumed by the V2 Templating Engine.
```

Comment:
Yeah this one is tricky.  These are the "templates" used during `pdk new ...`, and strictly speaking they're quite different (workflow wise) from the PDK templates that this RFC is addressing.

However....saying that... This RFC _could_ include these.  Currently the `template.yaml` has a `type` setting. I could introduce a new type called (for example) `new_objects`.  And these new template types could house those template files. This would still allow a graceful fallback to V1 `object_templates` if the directory existed.

Reply:
Yeah, I think the use case and application of the base template and object templates (that is, `pdk new module` versus `pdk new class`) are distinct enough that we can address them as separate things if needed.

Conceptually, the `pdk new class` templates are closer to `moduleroot_init` templates where they get generated once and then are unmanaged going forward.

Reply:
I'm marking this is Out-Of-Scope for this particular RFC. It is expected that later changes will address this